Addons/tsgExoticQuadrature.cpp

```diff
@@ -0,0 +1,183 @@
+ * DISPOSAL OF THE SOFTWARE.
+ */
+
+// TODO: Add better documentation.
+
+#ifndef __TASMANIAN_ADDONS_EXOTICQUADRATURE_CPP
+#define __TASMANIAN_ADDONS_EXOTICQUADRATURE_CPP
+
+#include "tsgCoreOneDimensional.hpp"
+#include "tsgLinearSolvers.hpp"
+#include <cassert>
+
+namespace TasGrid {
+
+// Evaluates a polynomial with roots given by \b roots at the point \b x.
+double poly_eval(const std::vector<double> &roots, double x) {
+    double eval_value = 1.0;
+    for (size_t i=0; i<roots.size(); i++) {
+        eval_value *= (x - roots[i]);
+    }
+    return eval_value;
+}
+// Generate roots.
+std::vector<std::vector<double>> getRoots(
+      const int n,
+      const std::vector<double> &ref_integral_weights,
+      const std::vector<double> &ref_points) {
+
+    // Compute the roots incrementally.
+    assert(ref_points.size() == ref_integral_weights.size());
+    std::vector<double>
+            poly_m1_vals(ref_points.size(), 0.0),
+            poly_vals(ref_points.size(), 1.0);
+    std::vector<double> alpha, beta;
+    std::vector<std::vector<double>> roots(n);
+    for (int i=0; i<n; i++) {
+        // Form the tridiagonal vectors, alpha and beta, of the Jacobi matrix.
+        double alpha_numr = 0.0;
+        double alpha_denm = 0.0;
+        double beta_numr = 0.0;
+        double beta_denm = 0.0;
+        for (int j=0; j<ref_points.size(); j++) {
+            alpha_numr += ref_points[j] * (poly_vals[j] * poly_vals[j]) * ref_integral_weights[j];
+            alpha_denm += (poly_vals[j] * poly_vals[j]) * ref_integral_weights[j];
+            beta_numr += (poly_vals[j] * poly_vals[j]) * ref_integral_weights[j];
+            beta_denm +=  (poly_m1_vals[j] * poly_m1_vals[j])  * ref_integral_weights[j];
+        }
+        alpha.push_back(alpha_numr / alpha_denm);
+        if (i >= 1) {
+            beta.push_back(beta_numr / beta_denm);
+        }
+        // Compute the roots. alpha and beta need to be copied because the
+        // tridiagonal eigensolver generates the eigenvalues in place and
+        // destroys some of its inputs.
+        roots[i] = alpha;
+        std::vector<double> offdiag(n-1);
+        for (int k=0; k<i; k++) {
+            offdiag[k] = sqrt(beta[k]);
+        }
+        TasmanianTridiagonalSolver::getSymmetricEigenvalues(i+1, roots[i], offdiag);
+        if (roots[i].size() % 2 == 1) {
+            roots[i][(roots[i].size() - 1) / 2] = 0.0; // Zero out the center for stability.
+        }
+        // Update the values of the polynomials at ref_points.
+        poly_m1_vals = poly_vals;
+        std::transform(ref_points.begin(), ref_points.end(), poly_vals.begin(),
+                       [&roots, i](double x){return poly_eval(roots[i], x);});
+    }
+    return roots;
+}
+
+//! \brief Evaluates a Lagrange basis polynomial at a point \b x.
+double lagrange_eval(int idx, const std::vector<double> roots, double x) {
+    double eval_value = 1.0;
+    for (size_t i=0; i<roots.size(); i++) {
+        if (i != idx) {
```

Comment:
if -statement in a for-loop can be expensive.

Break into two for-loops going to `0 ... idx-1` and `idx+1 ... roots.size()`